src/helpers.rs

```diff
@@ -0,0 +1,121 @@
+//! Helper methods for implementing the `ff` traits.
+
+use subtle::{Choice, ConditionallySelectable, ConstantTimeEq, CtOption};
+
+use crate::PrimeField;
+
+/// Constant-time implementation of Tonelli–Shanks' square-root algorithm for
+/// `p mod 16 = 1`.
+///
+/// https://eprint.iacr.org/2012/685.pdf (page 12, algorithm 5)
```

Comment:
```suggestion
/// <https://eprint.iacr.org/2012/685.pdf> (page 12, algorithm 5)
/// Steps 2-5 of the algorithm are omitted because they are only
/// needed to detect non-square input; it is more efficient to do
/// that by checking at the end whether the square of the result
/// is the input.
```

Reply:
Is this useful in the method documentation, or should this just be internal documentation? I think the latter.

Reply:
Since the implemented algorithm is somewhat different from the one in the paper, the reference probably shouldn't be in the public doc at all.

Reply:
Added new suggestions for documentation changes, which I will merge after your re-review (in case you have further suggestions).